presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushRemoteExchangeThroughGroupId.java

```diff
@@ -0,0 +1,123 @@
+
+public final class PushRemoteExchangeThroughGroupId
+        implements Rule<ExchangeNode>
+{
+    private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+    private static final Pattern<ExchangeNode> PATTERN = exchange()
+            .matching(exchange -> exchange.getScope().isRemote())
+            .matching(exchange -> exchange.getType() != REPLICATE)
+            .with(source().matching(
+                    groupId().capturedAs(GROUP_ID)));
+    @Override
+    public Pattern<ExchangeNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return SystemSessionProperties.shouldPushRemoteExchangeThroughGroupId(session);
+    }
+
+    @Override
+    public Result apply(ExchangeNode node, Captures captures, Context context)
+    {
+        GroupIdNode groupIdNode = captures.get(GROUP_ID);
+        Set<VariableReferenceExpression> commonGroupingColumns = groupIdNode.getCommonGroupingColumns();
+        if (commonGroupingColumns.isEmpty()) {
+            // If there is no common grouping column we can not push exchange operator below groupId.
+            return Result.empty();
+        }
+
+        PartitioningScheme partitioningScheme = node.getPartitioningScheme();
+        List<VariableReferenceExpression> inputs = getOnlyElement(node.getInputs());
+        inputs = removeVariable(inputs, groupIdNode.getGroupIdVariable());
+        inputs = replaceAlias(inputs, groupIdNode.getGroupingColumns());
+
+        List<VariableReferenceExpression> outputLayout = partitioningScheme.getOutputLayout();
+        outputLayout = removeVariable(outputLayout, groupIdNode.getGroupIdVariable());
+        outputLayout = replaceAlias(outputLayout, groupIdNode.getGroupingColumns());
+
+        List<VariableReferenceExpression> partitionColumns = replaceAlias(new ArrayList<>(commonGroupingColumns), groupIdNode.getGroupingColumns());
+
+        return Result.ofPlanNode(new GroupIdNode(
+                node.getSourceLocation(),
+                groupIdNode.getId(),
+                new ExchangeNode(
+                        node.getSourceLocation(),
+                        node.getId(),
+                        node.getType(),
+                        node.getScope(),
+                        new PartitioningScheme(
+                                Partitioning.create(partitioningScheme.getPartitioning().getHandle(), partitionColumns),
+                                outputLayout,
+                                partitioningScheme.getHashColumn(),
+                                partitioningScheme.isReplicateNullsAndAny(),
+                                partitioningScheme.getBucketToPartition()),
+                        ImmutableList.of(groupIdNode.getSource()),
+                        ImmutableList.of(inputs),
+                        node.isEnsureSourceOrdering(),
+                        Optional.empty()),
```

Comment:
Shouldn't this be node.getOrderingScheme()?

Reply:
updated